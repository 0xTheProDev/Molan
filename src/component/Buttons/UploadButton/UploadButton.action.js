// @flow
import {
  InputElementRefObject,
  FileInputChangeHandler,
  FileInputErrorHandler,
} from './UploadButton.flow';

export function onFileButtonClick(refObject: InputElementRefObject): void {
  return function onClickHandler(event: MouseEvent) {
    if (refObject.current) {
      refObject.current.click();
    }
  }
}

export function onFileInputChange(onUploadFile: FileInputChangeHandler, onUploadFileError: FileInputErrorHandler) {
  return function onChangeHandler(event: InputEvent): void {
    const files: FileList = event.target.files;

    if (!files || files.length !== 1) {
      onUploadFileError();
      return;
    }

    const fileReader = new FileReader();

    fileReader.addEventListener('load', (ev: ProgressEvent<FileReader>) => {
      onUploadFile(ev.target.result);
    });
    fileReader.addEventListener('error', (ev: ProgressEvent<FileReader>) => {
      onUploadFileError(ev.target.error);
    });

    fileReader.readAsText(files[0], 'utf-8');
  };
}
