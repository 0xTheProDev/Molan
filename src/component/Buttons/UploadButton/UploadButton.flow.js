// @flow
export type UploadButtonProps = {
  UploadButtonText: string,
  onUploadFile: FileInputChangeHandler,
  onUploadFileError: FileInputErrorHandler,
};

export type UploadButtonInternalProps = {
  AllowedFileTypes: string,
  InputElementRef: InputElementRefObject,
  UploadButtonText: string,
  onChangeHandler: (event: InputEvent) => void,
  onClickHandler: (event: MouseEvent) => void,
};

export type InputElementRefObject = {
  current?: HTMLInputElement,
};

export type FileInputChangeHandler = (fileContent: string) => void;

export type FileInputErrorHandler = (error?: string) => void;
