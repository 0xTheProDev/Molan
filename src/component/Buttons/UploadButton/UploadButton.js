// @flow
import React from 'react';
import { Button, Icon } from 'semantic-ui-react';
import { noop } from 'util/common';

import { UploadButtonInternalProps } from './UploadButton.flow';
import Styles from './UploadButton.module.css';

export function UploadButton(props: UploadButtonInternalProps) {
  const {
    InputElementRef,
    UploadButtonText,
    AllowedFileTypes,
    onClickHandler,
    onChangeHandler,
  } = props;

  return (
    <Button
      animated="vertical"
      floated="right"
      title={UploadButtonText}
      onClick={onClickHandler}
    >
      <Button.Content hidden>{UploadButtonText}</Button.Content>
      <Button.Content visible>
        <Icon name="upload" />
      </Button.Content>
      <input
        ref={InputElementRef}
        type="file"
        accepts={AllowedFileTypes}
        onChange={onChangeHandler}
        className={Styles['file-input']}
      />
    </Button>
  );
}

UploadButton.defaultProps = {
  UploadButtonText: 'Upload',
  onChangeHandler: noop,
  onClickHandler: noop,
};
