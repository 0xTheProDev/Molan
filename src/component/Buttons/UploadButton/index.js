// @flow
import React from 'react';
import { MIME_TYPES } from 'constants/templates';

import { UploadButton } from './UploadButton';
import { UploadButtonProps } from './UploadButton.flow';
import { onFileButtonClick, onFileInputChange } from './UploadButton.action';

export default function UploadButtonFactory(props: UploadButtonProps) {
  const {
    UploadButtonText,
    onUploadFile,
    onUploadFileError,
  } = props;

  const AllowedMimeTypes = new Set(Object.values(MIME_TYPES));
  const AllowedFileTypes = Array.from(AllowedMimeTypes).join(',');

  const InputElementRef = React.createRef();

  const onClickHandler = onFileButtonClick(InputElementRef);
  const onChangeHandler = onFileInputChange(onUploadFile, onUploadFileError);

  return (
    <UploadButton
      AllowedFileTypes={AllowedFileTypes}
      InputElementRef={InputElementRef}
      UploadButtonText={UploadButtonText}
      onClickHandler={onClickHandler}
      onChangeHandler={onChangeHandler}
    />
  );
}

export { UploadButtonProps, UploadButtonFactory as UploadButton };
