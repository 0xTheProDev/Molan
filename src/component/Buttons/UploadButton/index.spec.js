import React from 'react';
import { Button } from 'semantic-ui-react';

import UploadButton from './index';

describe('Component::Buttons:UploadButton::Integration Suite', () => {
  let
    onUploadFileStub,
    onUploadFileErrorStub,
    minProps;

  beforeAll(() => {
    onUploadFileStub = jest.fn();
    onUploadFileErrorStub = jest.fn();

    minProps = {
      UploadButtonText: 'Upload',
      onUploadFile: onUploadFileStub,
      onUploadFileError: onUploadFileErrorStub,
    }; 
  });

  it('should render with base structure', () => {
    const wrapper = mount(<UploadButton {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should let user upload file when clicked on Upload button', () => {
    const
      eventListeners  = {},
      FileContent     = 'Text Content',
      ErrorMessage    = 'Some Error Message';
    
    const File = new Blob([ FileContent ], { type: 'text/plain' });

    const
      HTMLInputElementClickStub       = jest.spyOn(HTMLInputElement.prototype, 'click'),
      FileReaderAddEventListenerStub  = jest.spyOn(FileReader.prototype, 'addEventListener'),
      FileReaderReadAsTextStub        = jest.spyOn(FileReader.prototype, 'readAsText');

    FileReaderAddEventListenerStub.mockImplementation((eventName, eventHandler) => {
      eventListeners[eventName] = eventHandler;
    });

    const wrapper = mount(<UploadButton {...minProps} />);

    /** User clicks on Upload button and Explorer window pops-up */
    wrapper.find(Button).simulate('click');
    expect(HTMLInputElementClickStub).toHaveBeenCalled();

    /** User selects a file */
    wrapper.find('input').simulate('change', {
      target: {
        files: [ File  ],
      },
    });
    expect(FileReaderReadAsTextStub).toHaveBeenCalledWith(File, 'utf-8');

    /** Should invoke callback after file content being loaded  */
    const loadEventListener = eventListeners.load;
    loadEventListener({
      target: {
        result: FileContent,
      },
    });
    expect(onUploadFileStub).toHaveBeenCalledWith(FileContent);

    /** Should invoke callback if fails to load file content */
    const errorEventListener = eventListeners.error;
    errorEventListener({
      target: {
        error: ErrorMessage,
      },
    });
    expect(onUploadFileErrorStub).toHaveBeenCalledWith(ErrorMessage);
  });
});
