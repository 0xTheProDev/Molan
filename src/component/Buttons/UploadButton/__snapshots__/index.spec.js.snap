// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component::Buttons:UploadButton::Integration Suite should render with base structure 1`] = `
<UploadButtonFactory
  UploadButtonText="Upload"
  onUploadFile={[MockFunction]}
  onUploadFileError={[MockFunction]}
>
  <UploadButton
    AllowedFileTypes="text/x-csrc,text/x-c++src,text/x-java,text/x-python,text/plain"
    InputElementRef={
      Object {
        "current": <input
          accepts="text/x-csrc,text/x-c++src,text/x-java,text/x-python,text/plain"
          class="file-input"
          type="file"
        />,
      }
    }
    UploadButtonText="Upload"
    onChangeHandler={[Function]}
    onClickHandler={[Function]}
  >
    <Button
      animated="vertical"
      as="button"
      floated="right"
      onClick={[Function]}
      title="Upload"
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui vertical animated right floated button"
              title="Upload"
            >
              <div
                class="hidden content"
              >
                Upload
              </div>
              <div
                class="visible content"
              >
                <i
                  aria-hidden="true"
                  class="upload icon"
                />
              </div>
              <input
                accepts="text/x-csrc,text/x-c++src,text/x-java,text/x-python,text/plain"
                class="file-input"
                type="file"
              />
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui vertical animated right floated button"
                title="Upload"
              >
                <div
                  class="hidden content"
                >
                  Upload
                </div>
                <div
                  class="visible content"
                >
                  <i
                    aria-hidden="true"
                    class="upload icon"
                  />
                </div>
                <input
                  accepts="text/x-csrc,text/x-c++src,text/x-java,text/x-python,text/plain"
                  class="file-input"
                  type="file"
                />
              </button>,
            }
          }
        >
          <button
            className="ui vertical animated right floated button"
            onClick={[Function]}
            title="Upload"
          >
            <ButtonContent
              hidden={true}
            >
              <div
                className="hidden content"
              >
                Upload
              </div>
            </ButtonContent>
            <ButtonContent
              visible={true}
            >
              <div
                className="visible content"
              >
                <Icon
                  as="i"
                  name="upload"
                >
                  <i
                    aria-hidden="true"
                    className="upload icon"
                    onClick={[Function]}
                  />
                </Icon>
              </div>
            </ButtonContent>
            <input
              accepts="text/x-csrc,text/x-c++src,text/x-java,text/x-python,text/plain"
              className="file-input"
              onChange={[Function]}
              type="file"
            />
          </button>
        </RefFindNode>
      </Ref>
    </Button>
  </UploadButton>
</UploadButtonFactory>
`;
